### PR TITLE
[17.0][IMP] base_tier_validation: Merge module with base_tier_validation_waiting

### PR DESCRIPTION
**Pull request by bosd**
As proposed as an idea in:
https://github.com/OCA/server-ux/pull/787#issuecomment-1871461565
This is the PR for the migration of base_tier_validation_waiting into the base module `base_tier_validation_waiting`.

Please fasstrack, as v17 is not deployed yet in production, this is the best time to introduce this refactor.

To the contributors/authers/reviewers of the migration pr..

@celm1990, @nguyenminhchien, @bofiltd, @moitabenfdz, @sonhd91, @Dranyel-Bosd, @BT-jcolmeiro
Can you please review?

### CODE REVIEW

**Comment by bosd** [comment]
Some background info...

Since sequential validations are supported in the base module. It makes sense from an user perspective to only sent out notifications to the validator if it is thier turn in the sequence to review.
Merging of these two modules accomplish that.

Before this pr (and without the waiting module) it was only possible to sent out and notification when a record was created. A reviewer in the end of the sequence got the e-mail notification before it's turn.
If the user decided to go to the document and make an attempt to do so. He was unable to do it.. Because it was not their turn yet..
Since there was no capability to get a notification. The user- had to visit the record now and then to see if it was their turn to validate.
This was a poor flow from usabulity point of view.
Easily fixed by the waiting module.

The merge of into the base module. Should make it more clear to users/functional consultants. As well as easier to maintain.

**Review by bofiltd — APPROVED** [review]
Makes sense.
Functional test locally. LGTM :champagne:

File: base_tier_validation/models/tier_review.py
Comment (celm1990) [inline base_tier_validation/models/tier_review.py:95]: I'm uncertain whether it's advisable to update fields within a `computed function` if those fields are not explicitly defined as computed (field `status`).

File: base_tier_validation/models/tier_validation.py
Comment (celm1990) [inline base_tier_validation/models/tier_validation.py]:
```suggestion
                and self.env.user in r.reviewer_ids
```

File: base_tier_validation/models/tier_validation.py
Comment (celm1990) [inline base_tier_validation/models/tier_validation.py]: Are parentheses valid?
```suggestion
        waiting_reviews = tier_reviews.filtered(
            lambda r: r.status == "waiting"
            or r.approve_sequence_bypass
            and self.env.user in r.reviewer_ids
        )
```

File: base_tier_validation/models/tier_validation.py
Comment (celm1990) [inline base_tier_validation/models/tier_validation.py]:
```suggestion
            and self.env.user in r.reviewer_ids
```

File: base_tier_validation/models/tier_validation.py
Comment (celm1990) [inline base_tier_validation/models/tier_validation.py]: If at this point in the code, it is certain that this method exists because it is not called directly?, it enhances readability.
```suggestion
                rec.message_subscribe(
                    partner_ids=users_to_notify.mapped("partner_id").ids
                )
                rec.message_post(
                    subtype_xmlid=self._get_requested_notification_subtype(),
                    body=rec._notify_requested_review_body(),
                )
```

**Review by celm1990 — COMMENTED** [review]
first code review; I will test it for functionality.

**Review by Dranyel-Bosd — APPROVED** [review]
Functional test locally, LGTM 🍾

**Comment by celm1990** [comment]
I tested these changes, but I have some doubts. For example, I have three reviews for three specific users:

R1 -> Admin
R2 -> U2
R3 -> U3
image


**Before this commit:**
When Admin approved R1, an email is sent to U2 and U3 notifying them that R1 is approved.

**After this commit:**
When Admin approves R1, an email is sent to both U2 and U3, notifying them that R1 is approved(as previous). Additionally, a new email is sent to U2 and **U3** informing them that a new review (R2) has been created. However, for the purpose of this changes, if the intention is to notify only the next user (**U2**), I hope that only **U2** is notified and that no email is sent to **U3**. This is because if U3 receives an email, they might attempt to review and approve, causing confusion as it is not their turn yet. we would revert back to the initial problem.

image

Please let me know if I understood the functionality of the module correctly.

**Comment by bosd** [comment]
> Additionally, a new email is sent to U2 and **U3** informing them that a new review (R2) has been created. However, for the purpose of this changes, if the intention is to notify only the next user (**U2**), I hope that only **U2** is notified and that no email is sent to **U3**. This is because if U3 receives an email, they might attempt to review and approve, causing confusion as it is not their turn yet. we would revert back to the initial problem.

Yes, I think you understand the functionality correctly.
The main problem before, was that U3 got an invitation email to review. 
But he coul'dnt review because it was not his turn yet.
Before this PR, he was unable to get a notification when it actually was his turn.

Another use case is that U3 could be upper management. And only wants to be bothered with messges which made it past the previous tiers. 

I'm under the impression you have the functionality and use case correct.
Yet, I'm unable to determine from your feedback if the module passed the functional test.
Dunno if it is the case, but the notification settings might be a bit "misleading"/unclear.
In a sequential flow, I would reccomend to only check the `Notify Reviewers on reaching Pending` box.
Ignoring all the other notifications, this should happen.


```mermaid
flowchart TD
    A[U0 validation request] -->|Notification message: request to review sent to U1| B[U1 Approves SO]
    B -->|Notification message: request to review sent to  U2| C[U2 approves SO]
    C -->|Notification message: request to review sent to U3| D[U3 approves SO]
    D--> E(The End\n Record = validated)
```

**Comment by bosd** [comment]
While out of scope of this pr, we might want to look into who is receiving the "approved messages".
To have it go to the original document owner makes sense. In case U3 is upper management, we don't want to bother him with all the back and forth happening in the tiers before.

**Review by celm1990 — APPROVED** [review]
Thanks for clarifying. I unchecked "Notify on Creation" as per your suggestion to reduce emails. However, I noticed that other users are still being notified when reviews are requested. I suspect this is happening because of followers. So, LGTM .

image

By the way, in the other PR(Improve UI), you can add these fields as `optional="show"`  😅

NOTE: Please let me know your comments regarding my comment on computed fields not being defined as computed explicitly. Could this be causing cache inconsistency?

File: base_tier_validation/models/tier_review.py
Comment (bosd) [reply base_tier_validation/models/tier_review.py:95]: >NOTE: Please let me know your comments regarding my comment on computed fields not being defined as computed explicitly. Could this be causing cache inconsistency?

tbh, I'm a bit overasked by this.. 
Maybe it is not the ideal situation. (yet, I'm unaware of any alternative method to implement this).
Well, we've deployed this code succesfully in production. W/o any problems. 
Judging from the experience it is okay..
But happy to learn from the more experienced dev's.

**Comment by bosd** [comment]
Oopsie, I answered on the comment above, but it does'nt show up down here.
If more improvements are suggested I'm happy to include them.
But as other mirgrations depend on this base module, ideally we would move forward.

@OCA/tools-maintainers I think this one is ready for merge..

**Comment by celm1990** [comment]
@LoisRForgeFlow @pedrobaeza, could you take a look at this PR? What do you think, is it possible to merge?

**Comment by bosd** [comment]
@kittiu Can you please review as well?

**Comment by bosd** [comment]
@gurneyalex Can you take a look as well? 🙏

**Comment by bosd** [comment]
@oca/edi-maintainers Can we please fasttrack this one? :pray:

**Comment by bosd** [comment]
Oops, used the wrong handle to fasttrack this one.. [@OCA/server-ux-maintainers](https://github.com/orgs/OCA/teams/server-ux-maintainers) seems to be the correct one?

Or @rousseldenis Do you have super powers here as well?

**Comment by rousseldenis** [comment]
> Oops, used the wrong handle to fasttrack this one.. [@OCA/server-ux-maintainers](https://github.com/orgs/OCA/teams/server-ux-maintainers) seems to be the correct one?
> 
> Or @rousseldenis Do you have super powers here as well?

Nope.

**Review by LoisRForgeFlow — COMMENTED** [review]
Functional review, in general I think this is a great proposal, thanks! Some comments below:

1. This field seems to be a bit misplaced:

image

2. Time to add the sequence handler widget?

image

3. There are multiple notifications with the same message. This one is not a blocking comment as I know this was an issue before this change too, but with this sequencial approval it becomes more noisy.

image

**Comment by bosd** [comment]
> Functional review, in general I think this is a great proposal, thanks! Some comments below:
> 
> 1. This field seems to be a bit misplaced:
> 
> image
>

The UI Improvements are made in follow up pr.
https://github.com/OCA/server-ux/pull/804


> 2. Time to add the sequence handler widget?
> 
> image
> 
Great Suggestion.. 
Will add it, In this PR or the UI PR. In case this one gets merged first.

> 3. There are multiple notifications with the same message. This one is not a blocking comment as I know this was an issue before this change too, but with this sequencial approval it becomes more noisy.
> 
> image

I agree, The messages and the content could be improved.
Have been thinking about rewording the message to.. But decided to not do big chances for now.
Adding the reviewer name to which the review is requested from could be usefull.
But will also lead to confusion by request from a group or server action.


@LoisRForgeFlow Are you ok to merge this one as is?
So the other modules which depend on it can be refactored during the migration.

**Comment by LoisRForgeFlow** [comment]
@bosd Yes, I think we can merge as is, and continue the improvements on followup PRs. I don't want to hold this because of my lack of time to review

File: base_tier_validation/models/tier_review.py
Comment (StefanRijnhart) [inline base_tier_validation/models/tier_review.py:80]: @bosd Can you help me understand `can_review` vs. `waiting` and `pending`? Is it always the case that `can_review` is true if and only if its status is `pending`? Would it be safe for me to propose a simplification based on this? Or am I missing something?

Comment (bosd) [reply base_tier_validation/models/tier_review.py:80]: Oof, thats been a while.
Doing this without code check, out of memory:
can_review_ everyone who can review the record.
Waiting = when it is not yet the users turn to approve
Pending = his turn to approve in sequential flow?

Comment (StefanRijnhart) [reply base_tier_validation/models/tier_review.py:80]: Thanks for checking in on this! It seems that can_review on tier.validation is indeed user aware, but can_review on tier.review is not. That's why I'm thinking it *is* simply the case that can_review equals 'pending' so I think the related code can be simplified, or maybe `can_review` on the tier.review level can be dropped entirely.

**Pull request merged**
